### ISSUE TITLE
Flax scan split_rng does not work when called in functional form (modular form works)

### ISSUE THREAD

**Issue by bondquant** [google/flax#4950]
Sirs,

  It seems not possible to split RNGs in "functional-form" Flax scan (that is when it's calling a function directly instead of separating the code to a Module). It affects in particular any Dropout modules you happen to call directly in the target function (by essentially freezing randomness across time). Modules `FlaxScanModule` and `FlaxScanCallingModule` below work just fine, however `FlaxScanFunction` is unable to split its RNGs. Is this expected behaviour or a bug?

Many thanks,

  Alexey

```
from flax import linen as nn
from jax import random as jrnd, numpy as jnp


class Foo(nn.RNNCellBase):
    def __call__(self, carry, *args, **kwargs):
        return carry, jrnd.key_data(self.make_rng('dropout'))


class FlaxScanModule(nn.Module):
    split_rngs: bool

    @nn.compact
    def __call__(self, x):
        scan = nn.scan(Foo, split_rngs={'dropout': self.split_rngs})

        _, x = scan()((), x)

        return x


class FlaxScanCallingModule(nn.Module):
    split_rngs: bool

    @nn.compact
    def __call__(self, x):
        def body_fn(cell, carry, x):
            return cell(carry, x)

        scan = nn.scan(body_fn, split_rngs={'dropout': self.split_rngs})

        _, x = scan(Foo(), (), x)

        return x


class FlaxScanFunction(nn.Module):
    split_rngs: bool

    @nn.compact
    def __call__(self, x):
        def body_fn(_cell, carry, _x):
            # Doesn't seem possible to split the key in direct call to `make_rng`.
            # Note that this is what happens if you happen to have any `Dropout`
            # modules in this function. Basically, one must only use `nn.scan` to just
            # call `cell(carry, x)` without any further processing (cf. `FlaxScanCallingModule`).
            y = jrnd.key_data(self.make_rng('dropout'))
            return carry, y

        scan = nn.scan(body_fn, split_rngs={'dropout': self.split_rngs})

        _, x = scan(Foo(), (), x)

        return x


rngs = {'dropout': jrnd.key(1729)}
xs = jnp.ones((2,))

module_split = FlaxScanModule(split_rngs=True)
module_nosplit = FlaxScanModule(split_rngs=False)
module_output_split, _ = module_split.init_with_output(rngs, xs)
module_output_nosplit, _ = module_nosplit.init_with_output(rngs, xs)

# this works as expected
print(f'Module split: {module_output_split}') 
# [[2980801083 1655321879], [4258708554 3550203198]]

print(f'Module no_split: {module_output_nosplit}')
# [[4027286730   61492345], [4027286730   61492345]]

hybrid_split = FlaxScanCallingModule(split_rngs=True)
hybrid_nosplit = FlaxScanCallingModule(split_rngs=False)
hybrid_output_split, _ = hybrid_split.init_with_output(rngs, xs)
hybrid_output_nosplit, _ = hybrid_nosplit.init_with_output(rngs, xs)

# this too
print(f'Hybrid split: {hybrid_output_split}')
# [[3576790510 2841913155], [2367479996 2541703867]]

print(f'Hybrid no_split: {hybrid_output_nosplit}')
# [[ 612423757 1656726705], [ 612423757 1656726705]]

func_split = FlaxScanFunction(split_rngs=True)
func_nosplit = FlaxScanFunction(split_rngs=False)
func_output_split, _ = func_split.init_with_output(rngs, xs)
func_output_nosplit, _ = func_nosplit.init_with_output(rngs, xs)

# but this doesn't work, the keys are constant
print(f'Func split: {func_output_split}')
# [[4142386227 1365206574], [4142386227 1365206574]]

print(f'Func no_split: {func_output_nosplit}')
# [[4142386227 1365206574], [4142386227 1365206574]]
```

System info:
```
jax:    0.5.3
jaxlib: 0.5.3
numpy:  2.0.2
python: 3.12.11 (main, Jun  4 2025, 08:56:18) [GCC 11.4.0]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='20301c1c3418', release='6.1.123+', version='#1 SMP PREEMPT_DYNAMIC Sun Mar 30 16:01:29 UTC 2025', machine='x86_64')
```

**Comment by vfdev-5** [comment]
@bondquant thanks for the report. I wonder whether you could use recent versions of jax (0.7.1) and flax 0.11.2 and `nnx` instead of `linen`? You can check the example of `nnx.scan` where the rng split should work: https://flax.readthedocs.io/en/latest/api_reference/flax.nnx/transforms.html#flax.nnx.scan

Meanwhile, I'll try to take a look at this problem.

**Comment by bondquant** [comment]
Thanks for looking into this. I upgraded Jax to 0.7.1 and Flax to 0.11.2 and it's the exactly same. I am aware of the new-gen NNX, however, have existing codebase written in Flax Linen which cannot be upgraded quickly.

**Comment by vfdev-5** [comment]
@bondquant I think this is how it should be used:
```python
from flax import linen as nn
from jax import random as jrnd, numpy as jnp

class FlaxScanFunction(nn.Module):
    split_rngs: bool

    @nn.compact
    def __call__(self, x):
        def body_fn(_cell, carry, _x):
            y = jrnd.key_data(_cell.make_rng('dropout'))  # <------ use `_cell` instead of captured `self`
            return carry, y

        scan = nn.scan(body_fn, split_rngs={'dropout': self.split_rngs})
        _, x = scan(self, (), x)
        return x

rngs = {'dropout': jrnd.key(1729)}
xs = jnp.ones((2,))

func_split = FlaxScanFunction(split_rngs=True)
func_nosplit = FlaxScanFunction(split_rngs=False)
func_output_split, _ = func_split.init_with_output(rngs, xs)
func_output_nosplit, _ = func_nosplit.init_with_output(rngs, xs)

print(f'Func split: {func_output_split}')
# Func split: [[3605546469 2573901760]
#  [3631766359 1833908426]]
print(f'Func no_split: {func_output_nosplit}')
# Func no_split: [[4142386227 1365206574]
#  [4142386227 1365206574]]
```

**Comment by vfdev-5** [comment]
@bondquant I close the issue as resolved, please feel free to reopen if you think we need to discuss more this topic
